include/pybind11/numpy.h

```diff
@@ -316,7 +316,64 @@ class dtype : public object {
     }
 };
 
-class array : public buffer {
+namespace detail {
+    inline void fail_dim_check(size_t dim, size_t ndim, const std::string& msg) {
```

Comment:
I think this would be better as ```[[noreturn]] PYBIND11_NOINLINE inline void fail_dim_check(...```

(Also, unindent this part; I seem to recall @wjakob mentioning that not indenting functions in namespaces was one of his primary motivations for using the namespace macros).